File: 3240-minimum-number-of-flips-to-make-binary-grid-palindromic-ii/3240-minimum-number-of-flips-to-make-binary-grid-palindromic-ii.java
class Solution {
    public int minFlips(int[][] grid) {
        
        int n=grid.length,m=grid[0].length;
        int min=0;
        
        if(n%2!=0 && m%2!=0 && grid[n/2][m/2]==1){min++;}
        
        for(int i=0;i<n/2;i++){
            
            for(int j=0;j<m/2;j++){
                
                int c1=0;
                int e1=grid[i][j];
                int e2=grid[n-i-1][j];
                int e3=grid[i][m-j-1];
                int e4=grid[n-i-1][m-j-1];
                
                if(e1==1){c1++;}
                if(e2==1){c1++;}
                if(e3==1){c1++;}
                if(e4==1){c1++;}
                
                if(c1>4-c1){min+=(4-c1);}
                else{min+=c1;}
                
                
            }
            
        }
        
        
        if(n%2==0 && m%2==0){return min;}
        
        int go=0,gm=0,gz=0;
        if(n%2!=0){
            int i=(n/2);
            int j=0,k=m-1;
            while(j<k){
                if(grid[i][j]==0 && grid[i][k]==0){gz++;}
                else if(grid[i][j]==1 && grid[i][k]==1){go++;}
                else{gm++;}
                j++;k--;
            }
        }
        
        if(m%2!=0){
            int col=(m/2);
            int j=0,k=n-1;
            while(j<k){
                if(grid[j][col]==0 && grid[k][col]==0){gz++;}
                else if(grid[j][col]==1 && grid[k][col]==1){go++;}
                else{gm++;}
                j++;k--;
            }
        }
        
        if(go*2%4==0 || gm!=0){return min+gm;}
        
        
        return min+2;
        
        
    }
}

File: 3240-minimum-number-of-flips-to-make-binary-grid-palindromic-ii/3240-minimum-number-of-flips-to-make-binary-grid-palindromic-ii.py
class Solution:
    def minFlips(self, grid: List[List[int]]) -> int:
        all_row, all_col = len(grid), len(grid[0])

        result = 0
        one, one_plus_or_minus_region = 0, 0
        for row in range(ceil(all_row / 2)):
            for col in range(ceil(all_col / 2)):
                a, b, c, d = grid[row][col], grid[all_row - row - 1][col], grid[all_row - row - 1][all_col - 1 - col], grid[row][all_col - 1 - col]

                if all_row - 1 - row != row and all_col - 1 - col != col:
                    s = a + b + c + d
                    result += min(4 - s, s - 0)
                elif all_row - 1 - row == row and all_col - 1 - col == col:
                    one += a  # max = 1
                elif all_row - 1 - row == row and all_col - 1 - col != col:
                    one += (a + c) # max = 2
                    if a + c == 1:
                        one_plus_or_minus_region += 1
                else:
                    one += (a + b) # max = 2
                    if a + b == 1:
                        one_plus_or_minus_region += 1

        result += one_plus_or_minus_region

        min_t_add = 4
        for t in range(one - one_plus_or_minus_region, one + one_plus_or_minus_region + 1, 2):
            min_t_add = min(min_t_add, t % 4)
        result += min_t_add

        return result